<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Introducing R and RStudio</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/computer/
-->
<meta name="google-site-verification" content="t-Wgs-RXt8xi94RSoFamxoD2P2h-Hh7M_t_iOuxFZUo" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to R & RStudio</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop-intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="r-01-introduction.html">Introducing R and RStudio</a>
    </li>
    <li>
      <a href="r-02-exploration-dplyr.html">Data Exploration with dplyr</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="r-02-exploration-dplyr.html">Data Exploration with dplyr (cont.)</a>
    </li>
    <li>
      <a href="r-04-bioconductor.html">Using Packages / Libraries</a>
    </li>
    <li>
      <a href="r-05-help.html">Getting Help with R</a>
    </li>
    <li>
      <a href="r-03-exploration-ggplot.html">Data Visualization with ggplot</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop-wrap-up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Introducing R and RStudio</h1>

</div>


<style type="text/css">

body, td {
   font-size: 16px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

</style>
<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="objectives" class="section level1">
<h1>Objectives</h1>
<ul>
<li>Introduce R and RStudio.</li>
<li>Learn how to read in a csv file.</li>
<li>Discuss object naming conventions and types.</li>
<li>Discuss functions and parameters.</li>
<li>Learn how to get help with functions.</li>
<li>Get comfortable with errors and asking for help.</li>
</ul>
</div>
<div id="introduction-to-r-and-rstudio" class="section level1">
<h1>Introduction to R and RStudio</h1>
<p>In these lessons we’ll use the <a
href="https://www.gapminder.org/">gapminder dataset</a> to explore the
relationship between a country’s life expectancy and the total value of
its finished goods and services, also known as the Gross Domestic
Product (GDP). To explore this relationship we need
<strong>data</strong> and a <strong>platform to analyze the
data</strong>.</p>
<p>We could explore the data with a spreadsheet program like Excel or
Google Sheets, but it would be cumbersome to record the steps used to
explore and make changes to the original data. Instead, we’ll use a
programming language to explore and summarize the data. In particular
we’ll use R and RStudio to create tabular summaries of the data as well
as plots.</p>
<p>We’ll use R because it is:</p>
<ul>
<li>Open source, meaning it’s free!</li>
<li>Widely used, meaning it has a large community of support to get
help.</li>
<li>Powerful, meaning it has many packages that extend and customize its
abilities.</li>
</ul>
<p>We’ll also use R because this is a precursor workshop to RNA-seq
Demystified, which will teach you how to use R for differential
expression analysis.</p>
<p>To get started, we’ll use <a
href="https://posit.co/products/open-source/rstudio/">RStudio</a>, an
integrated development environment (IDE). It acts as a graphical
interface to R that has many helpful features, as we’ll see. Note that R
and RStudio are different, but complimentary. You need R to use
RStudio.</p>
</div>
<div id="log-in-to-rstudio-server" class="section level1">
<h1>Log in to RStudio Server</h1>
<p>To get started, we’ll go to <a
href="http://bfx-workshop02.med.umich.edu"
target="blank">http://bfx-workshop02.med.umich.edu</a></p>
<p>The login page for the server looks like:</p>
<p><img src="images/01_rstudio_login_screen.png" alt="rstudio login screen" style="width: 1000px;"/></p>
<p>Enter your user credentials and click <kbd>Sign In</kbd>. The RStudio
interface should load and look like:</p>
<p><img src="images/02_rstudio_session_default.png" alt="rstudio default session" style="width:1000px;"/></p>
<p><strong>Checkpoint</strong></p>
</div>
<div id="orienting-on-rstudio" class="section level1">
<h1>Orienting on RStudio</h1>
<p>RStudio is an integrated development environment where you can write,
execute, and see the results of your code. The interface is arranged in
different panes:</p>
<ul>
<li>The Console pane along the left where you can enter commands and
execute them.</li>
<li>The Environment pane in the upper right shows any variables you have
created, along with their values.</li>
<li>The pane in the lower right has a few functions:
<ul>
<li>The Files tab let’s you navigate the file system.</li>
<li>The Plots tab displays any plots from code run in the Console.</li>
<li>The Help tab displays the documentation of functions.</li>
</ul></li>
</ul>
<div id="commands-in-the-console" class="section level2">
<h2>Commands in the Console</h2>
<p>We can input commands directly into the Console and press
<KBD>Enter</KBD> to execute them. Try it with:</p>
<pre><code>&gt; 2+2
[1] 4</code></pre>
<p>If we enter commands directly into the Console and execute them,
<strong>there is no record</strong> that persists after we close
RStudio.</p>
<p><img src="images/03_rstudio_console_execution.png" alt="rstudio console execution" style="width:1000px;"/></p>
<p><strong>Checkpoint</strong></p>
</div>
<div id="commands-in-the-script" class="section level2">
<h2>Commands in the Script</h2>
<p>To keep that record, we will enter our commands in a script. When
first opening RStudio, there is no pane to write code to save as a file.
But by clicking on the icon in the upper-right-most of the interface (a
blank piece of paper with a + sign), and selecting <KBD>R Script</KBD> a
new pane opens. This last pane is the Source (or Script) pane, and it is
here that we’ll keep a running record of the commands we learn in this
workshop.</p>
<p><img src="images/04_rstudio_create_script.png" alt="rstudio create script" style="width:1000px;"/></p>
<p>Which will swing open the Script pane:</p>
<p><img src="images/05_rstudio_script_pane.png" alt="rstudio script pane" style="width:1000px;"/></p>
<p>In the Script pane, enter:</p>
<pre><code>3+2</code></pre>
<p>Notice that if we press <KBD>Enter</KBD> in the Source pane, we get a
new line. <strong>The code does note execute.</strong> In order to
execute the code we press <KBD>Ctrl + Enter</KBD> <strong>on the line of
code we want to execute</strong>. We then see that code executed, along
with its result in in the Console pane.</p>
<p><img src="images/06_rstudio_script_execution.png" alt="rstudio script execution" style="width:1000px;"/></p>
<blockquote>
<h2 id="note-script-pane-and-console-pane">Note: Script Pane and Console
Pane</h2>
<p>The code written in the Script Pane is just text until it is executed
with <KBD>Control + Enter</KBD>. The Console Pane is the record of the
code that has been executed. <br></p>
</blockquote>
</div>
<div id="configuring-rstudio" class="section level2">
<h2>Configuring RStudio</h2>
<p>All of the panes in RStudio have configuration options. For example,
you can minimize/maximize a pane or resize panes by dragging the
borders. The most important customization options for pane layout are in
the <KBD>View</KBD> menu. Other options such as font sizes,
colors/themes, and more are in the <KBD>Tools</KBD> menu under
<KBD>Global Options</KBD>.</p>
<p>We can enable soft-wrapping of code by selecting <KBD>Code</KBD> and
then <KBD>Soft Wrap Long Lines</KBD>.</p>
<p><img src="images/07_rstudio_soft_wrap.png" alt="rstudio code execution" style="width:1000px;"/></p>
</div>
</div>
<div id="workshop-flow" class="section level1">
<h1>Workshop Flow</h1>
<p>The code we execute together is available in three forms:</p>
<ol style="list-style-type: decimal">
<li>By viewing the instructor’s Zoom screen share,</li>
<li>By code pasted into Slack, and</li>
<li>By code blocks on these web pages.</li>
</ol>
<p>As a learner you are welcome to get the same code into your Script
Pane by:</p>
<ol style="list-style-type: decimal">
<li>Typing the code from the Zoom screen share into your Script
Pane,</li>
<li>Copying the code from Slack and pasting into your Script Pane,
or</li>
<li>Copying the code from the code blocks on these web pages and pasting
into your Script Pane.</li>
</ol>
<p><strong>Questions?</strong></p>
</div>
<div id="creating-an-rstudio-project" class="section level1">
<h1>Creating an RStudio project</h1>
<p>We’ll now begin our examination of the gapminder dataset by creating
an <strong>RStudio Project</strong>. An RStudio project allows you to
more easily:</p>
<ul>
<li>Save data, files, variables, packages, etc. related to a specific
analysis project</li>
<li>Restart work where you left off</li>
<li>Collaborate, especially if you are using version control such as <a
href="http://swcarpentry.github.io/git-novice/">git</a>.</li>
</ul>
<ol style="list-style-type: decimal">
<li>To create a project, go to the <kbd>File</kbd> menu, and click
<kbd>New Project…</kbd>. If you are prompted to save your
<code>~/.Rhistory</code>, select No. The following window should
appear:</li>
</ol>
<center>
<img src="images/new_project_window.png" alt="rstudio default session" style="width: 600px;"/>
</center>
<ol start="2" style="list-style-type: decimal">
<li><p>In this window, select <kbd>Existing Directory</kbd>. For
“Project working directory”, click <kbd>Browse…</kbd>, select the “CF_R”
folder, and click <kbd>Choose</kbd>. This will use the
<code>/home/workshop/user/CF_R</code> folder as the project
directory.</p></li>
<li><p>Finally click <kbd>Create Project</kbd>. In the “Files” tab of
your output pane (more about the RStudio layout in a moment), you should
see an RStudio project file, <strong>CF_R.Rproj</strong>. All RStudio
projects end with the “<strong>.Rproj</strong>” file extension.</p></li>
</ol>
<p><strong>Checkpoint</strong></p>
<div id="creating-a-script" class="section level2">
<h2>Creating a script</h2>
<p>Let’s create an R script in this project to record the commands we’ll
learn:</p>
<ul>
<li>Click the <kbd>File</kbd> menu and select <kbd>New File</kbd> and
then <kbd>R Script</kbd>.</li>
<li>Before we go any further, save your script by clicking the save/disk
icon that is in the bar above the first line in the script editor, or
click the <kbd>File</kbd> menu and select <kbd>Save</kbd>.</li>
<li>In the “Save File” window that opens, select <kbd>New Folder</kbd>.
Name it <strong>“scripts”</strong>.</li>
<li>Finally, name your file <strong>“r_basics”</strong> in the “File
name” field.</li>
</ul>
<p>The new script <strong>r_basics.R</strong> is now in the
<code>scripts</code> folder. You can see that by clicking the
<code>scripts</code> folder in the “Files” pane. And you can go back up
to the main project folder by clicking the <code>..</code> to the right
of the up arrow in the “Files” pane. By convention, R scripts end with
the file extension <code>.R</code>.</p>
<p><strong>Checkpoint</strong></p>
</div>
</div>
<div id="loading-libraries" class="section level1">
<h1>Loading Libraries</h1>
<p>R has a number of functions that are present in the “base”
installation, but the power of R is that many libraries have been
written with myriad functionality. These libraries can be loaded with
the <code>library()</code> command. We’ll discuss how to download new
libraries later, but for now, let’s load the “tidyverse” package.</p>
<pre class="r"><code>library(tidyverse)</code></pre>
<blockquote>
<h2 id="note-package-loading-messages">Note: Package loading
messages</h2>
<p>When loading a package you may get lots of feedback. These aren’t
necessarily errors, but just give more information on the state of the
environment after loading the package. The <code>tidyverse</code> is
actually a collection of packages. The first section of the output
states which packages were lodaed and their versions. The second section
notes “Conflicts” that occur because the name of a function is used
multiple times. So <code>dplyr::filter() masks stats::filter()</code>
means that the <code>dplyr</code> library and the <code>stats</code>
library have functions called <code>filter()</code>, and that when
calling <code>filter()</code>, the <code>dplyr</code> version will be
the default. <br></p>
</blockquote>
<p><strong>Checkpoint</strong></p>
</div>
<div id="loading-a-csv-file" class="section level1">
<h1>Loading a csv file</h1>
<p>Let’s jump right in and load some of the gapminder data using the
<code>read_csv()</code> function:</p>
<pre class="r"><code>gapminder_1997 = read_csv(&#39;data/gapminder_1997.csv&#39;)</code></pre>
<pre><code>Rows: 142 Columns: 6
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
chr (2): country, continent
dbl (4): year, pop, lifeExp, gdpPercap

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Remember, with the cursor on this line we can click <KBD>Run</KBD>,
or we can type <KBD>Ctrl</KBD>+<KBD>Enter</KBD>. We should see some
output in the Console pane as well as <code>gapminder_1997</code> in the
Environment pane. We’ll explore the resulting data in later lessons.</p>
<p><strong>Checkpoint</strong></p>
<p>Let’s breakdown this command:</p>
<ul>
<li>We start with <code>gapminder_1997</code>, this is the name of the
<strong>object</strong> that will reference the data.</li>
<li>The <code>=</code> indicates the assignment operator. R also has the
assignment operator <code>&lt;-</code> but we’ll stick with
<code>=</code> in this workshop. The details about the distinction
aren’t relevant to our workshops.</li>
<li>The command ends with
<code>read_csv('data/gapminder_1997.csv')</code> which is a call to the
function <code>read_csv()</code> with an <strong>argument</strong>
stating the path of the file to read.</li>
</ul>
<p>The output of <code>read_csv()</code> in the Console pane gives us
useful information such as the dimensions of the data, the delimiter of
the file, and how the columns of the data were interpreted.</p>
</div>
<div id="basic-r-data-types" class="section level1">
<h1>Basic R data types</h1>
<p>We notice in the output of <code>read_csv()</code> that the
<code>country</code> and <code>continent</code> columns were intepreted
as character strings (<code>chr</code>) and the <code>year</code>,
<code>pop</code>, <code>lifeExp</code>, and <code>gdpPercap</code>
columns were interpreted as numbers (<code>dbl</code>).</p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Mode (abbreviation)</th>
<th>Type of data</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Numeric (num)</td>
<td>Numbers such as floating point/decimals (1.0, 0.5, 3.14), there are
also more specific numeric types (dbl - Double, int - Integer). These
differences are not relevant for most beginners and pertain to how these
values are stored in memory.</td>
</tr>
<tr class="even">
<td>Character (chr)</td>
<td>A sequence of letters/numbers in single ’ ’ or double ” ”
quotes.</td>
</tr>
<tr class="odd">
<td>Factor (fct)</td>
<td>A categorical variable with defined “levels” or categories.</td>
</tr>
<tr class="even">
<td>Logical</td>
<td>Boolean values - TRUE or FALSE.</td>
</tr>
</tbody>
</table>
</div>
<div id="assigning-values-to-objects" class="section level1">
<h1>Assigning values to objects</h1>
<p>Let’s assign values to some objects and observe each object after
assignment.</p>
<pre class="r"><code>name = &#39;Ben&#39;
name</code></pre>
<pre><code>[1] &quot;Ben&quot;</code></pre>
<pre class="r"><code>age = 26
age</code></pre>
<pre><code>[1] 26</code></pre>
<pre class="r"><code>name = &#39;Harry Potter&#39;
name</code></pre>
<pre><code>[1] &quot;Harry Potter&quot;</code></pre>
<p>Notice that with each assignment, the object appears in the
Environment pane. Also notice that by assigning <code>name</code> twice,
the value becomes the last assigned value, overwriting our initial
assignment. Also notice that if we evaluate the name of an object, it is
printed in the Console pane.</p>
<p>Let’s try to assign some more objects and see how things can go
wrong:</p>
<pre class="r"><code>1number = 3</code></pre>
<pre><code>Error: &lt;text&gt;:1:2: unexpected symbol
1: 1number
     ^</code></pre>
<pre class="r"><code>Flower = &#39;marigold&#39;
flower = &#39;rose&#39;</code></pre>
<pre class="r"><code>favorite number = 12</code></pre>
<pre><code>Error: &lt;text&gt;:1:10: unexpected symbol
1: favorite number
             ^</code></pre>
<p><strong>Checkpoint</strong></p>
<p>What happened here? We got “unexpected symbol” errors, which is R’s
obtuse way of saying it didn’t like the name of the object.</p>
<ul>
<li><code>1number</code> is problematic because object names can’t start
with numbers.</li>
<li><code>favorite number</code> is problematic because object names
can’t contain spaces.</li>
<li><code>Flower</code> and <code>flower</code> are distinct objects
because R is <strong>case-sensitive</strong>. Avoid variable names that
differ by case.</li>
</ul>
<blockquote>
<h2 id="guidelines-on-naming-objects">Guidelines on naming objects</h2>
<ul>
<li>Object names should be explicit and brief.</li>
<li>Object names cannot start with numbers or contain spaces.</li>
<li>Object names are case-sensitive.</li>
<li>Either separate words in object names like <code>object_name</code>
or <code>objectName</code>. <strong>Be consistent.</strong></li>
<li>Some names cannot be used because they are the names of fundamental
functions (e.g. if, else, for; see <a
href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/Reserved.html">here</a>
for complete list).</li>
</ul>
</blockquote>
</div>
<div id="calling-functions" class="section level1">
<h1>Calling functions</h1>
<p>Earlier we ran the code
<code>gapminder_1997 = read_csv('data/gapminder_1997.csv')</code>. As we
said before, <code>read_csv()</code> is a function and
<code>'data/gapminder_1997.csv'</code> is an argument to that function.
What happens if we just do:</p>
<pre class="r"><code>read_csv()</code></pre>
<pre><code>Error in read_csv(): argument &quot;file&quot; is missing, with no default</code></pre>
<p>We get an error in the Console pane. The key part of the message is
“argument ‘file’ is missing, with no default”. In other words, this
function needs to be told what to read because there is no default.</p>
<p>Not every function needs arguments, but many do. Try the following
functions:</p>
<pre class="r"><code>Sys.Date()</code></pre>
<pre><code>[1] &quot;2025-03-28&quot;</code></pre>
<pre class="r"><code>getwd()</code></pre>
<pre><code>[1] &quot;/home/workshop/rcavalca/workshop-intro-r-rstudio/source&quot;</code></pre>
<pre class="r"><code>round(3.1415, 2)</code></pre>
<pre><code>[1] 3.14</code></pre>
<p>Notice that we threw in <code>round()</code> which actually takes two
arguments. How could we have known that?</p>
<div id="getting-help-with-functions" class="section level2">
<h2>Getting help with functions</h2>
<p>When a function is unfamiliar, we’ll often look at the manual page
for the function to understand what arguments are required, what it
does, and what it outputs. By prepending a <code>?</code> in front of a
function name, you can access the manual page.</p>
<pre><code>?round</code></pre>
<p>The help page for <code>round()</code> tells us the function does
essentially what we’d expect, and gives some other related functions.
Note also that the arguments section gives us the names of the arguments
and what is expected of them. There is often a Details section to
describe nuances, and a Value section to describe the output. Finally,
there is an Examples section which gives examples of how to run the
code.</p>
</div>
<div id="arguments" class="section level2">
<h2>Arguments</h2>
<p>When we called <code>round(3.1415, 2)</code> it seemed like the first
argument is the thing we want to round, and the second argument is how
many digits we want. That tracks when we look at <code>?round</code>. R
can evaluate arguments of a function based on their
<strong>position</strong>, as we just saw. However, the preferred way to
call a function is to use the names of the arguments, as in:</p>
<pre class="r"><code>round(x = 3.14159, digits = 2)</code></pre>
<pre><code>[1] 3.14</code></pre>
<p>Calling a function, and using named arguments, increases the
readability of the code and reduces the chance of error, especially with
complex functions having many arguments.</p>
</div>
</div>
<div id="searching-for-functions" class="section level1">
<h1>Searching for functions</h1>
<p>Prepending a <code>?</code> in front a function name to find out more
about the function requires knowing the name of the function beforehand.
That won’t always be the case so there are a couple ways to search for R
functions.</p>
<ol style="list-style-type: decimal">
<li>Search the internet for “R function that does X”.</li>
<li>Use <code>help.search()</code>, as in
<code>help.search('Chi-squared test')</code></li>
</ol>
<p>Note that in the results of <code>help.search()</code> we see things
like, <code>stats::chisq.test</code>. Here the <code>::</code> is R
notation for <code>package_name::function</code>.</p>
<p><strong>Checkpoint</strong></p>
</div>
<div id="errosr-happn" class="section level1">
<h1>Errosr happn</h1>
<p>We already ran some code in this lesson that produced errors. We’re
going to run lots of code in this workshop and the next, and it’s likely
we’ll see more errors! The key steps to correcting an error are:</p>
<ol style="list-style-type: decimal">
<li>Read the error message to try to understand what might have gone
wrong. Note, not all error messages are clear, so don’t worry if the
error message doesn’t help you.</li>
<li>Immediately check the spelling of the command; <strong>most errors
are caused by typos</strong>.</li>
<li>Check that the objects going into the function are what’s
expected.</li>
</ol>
<p>If you’re still stuck as to why an error occurred (something we all
encounter), reach out for help. Make sure to post in Slack:</p>
<ol style="list-style-type: decimal">
<li>The exact command that caused the error.</li>
<li>The exact error message that resulted.</li>
</ol>
<p>This way we’ll more quickly be able to diagnose the problem.</p>
</div>
<div id="base-r-and-the-tidyverse" class="section level1">
<h1>Base R and the tidyverse</h1>
<p>If you’ve used R before, you may have learned commands that are
different what we’ll learn in this workshop. We’ll focus on functions
from the <code>tidyverse</code>, a collection of R packages designed to
work well together and and offer many features that aren’t part of a
fresh install of R (that is, “base R”). Generally the
<code>tidyverse</code> helps us write code that is easy to read and
maintain, as we’ll see.</p>
<p>The <code>tidyverse</code> is geared for data in the form of tables,
and it is <strong>very good</strong> at manipulating, summarizing, and
visualizing such data. However, data occurs in a variety of other shapes
and forms. In particular, in a bioinformatics context, the Bioconductor
repository of packages utilize data types that are not tables, and
therefore do not always work well with <code>tidyverse</code> functions.
We’ll see clearer examples of this in the RNA-seq Demystified workshop,
and you will undoubtedly encounter many examples in the future.</p>
<p>Some people ask “Should I learn tidyverse or base R?” and we think
that rather than either/or, it’s better to think of both/and. Knowing
base R and its approach will help in some contexts, while knowing
<code>tidyverse</code> will help in others.</p>
</div>
<div id="cheatsheets" class="section level1">
<h1>Cheatsheets!</h1>
<p>The <code>tidyverse</code> packages have excellent cheatsheets that
describe the functionality and usage of the packages. You can find them
in RStudio by going to the “Help” menu and selecting “Cheat Sheets”. The
two that will be most helpful in this workshop are “Data Visualization
with ggplot2” and “Data Transformation with dplyr”.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="bash-06-organization.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="r-02-exploration-dplyr.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
